src/mon/OSDMonitor.cc

```diff
@@ -3145,12 +3145,18 @@ bool OSDMonitor::check_failure(utime_t now, int target_osd, failure_info_t& fi)
   double my_grace = 0, peer_grace = 0;
   double decay_k = 0;
   if (g_conf()->mon_osd_adjust_heartbeat_grace) {
+    const osd_xinfo_t& xi = osdmap.get_xinfo(target_osd);
+    if (xi.down_stamp.sec()) {
+      int last_failure = now.sec() - xi.down_stamp.sec();
+      // Reset laggy parameters if failure interval exceeds a threshold.
+      if (grace_interval_threshold_exceeded(last_failure)) {
+        set_default_laggy_params(target_osd);
+      }
+    }
     double halflife = (double)g_conf()->mon_osd_laggy_halflife;
     decay_k = ::log(.5) / halflife;
```

Comment:
double-checked the math and the half-life decay should be effective, assuming the timestamps are accurate -- decay is 0.5 after one halflife, 0.25 after 2, etc. If the failure interval is inaccurate here, that would explain the extra grace period not going away. Are you still seeing more MOSDFailures sent after the osd is brought back in your testing?

Similarly, in the code setting the interval on boot, if the issue you're seeing with down_stamp not updating is present, the interval would always be set to mon_osd_max_laggy_interval, which would explain why rebooting was not helping the user who ran into this.

Your approach here of resetting things entirely is a nice fallback to avoid any further bugs like this.

Reply:
@jdurgin Apparently my previous fix was introduced in the wrong place. From my investigation and testing, the correct place to check for the threshold breach is when an osd is marked down
and its state is being updated/encoded within the pending incremental map.

When the osd is marked down, its state is set to CEPH_OSD_UP in
OSDMonitor::check_failure(). Eventually, when the incremental map is being encoded before applying it, a check is made to verify if the grace period is exceeding the set threshold if the
new state is set to CEPH_OSD_UP prior to resetting the laggy params. The down_stamp
for the osd is also updated as part of the reset.